# Notebook 2 – Data Collection & Integration

Project:
Predicting NFL Player Value to Identify Contract Overpayment Risk Using Machine Learning

Student:
William Reed

Course:
DSE6311 Capstone

Date:
July 2026

## Objective

The objective of this notebook is to collect, inspect, clean, and integrate player statistics, player information, and contract data into a unified master dataset. The resulting dataset will be used throughout the remainder of the project for exploratory analysis, feature engineering, and predictive modeling.

In [2]:
import polars as pl

In [3]:
import nflreadpy as nfl

dir(nfl)

['__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '__version__',
 'cache',
 'clear_cache',
 'config',
 'downloader',
 'get_current_season',
 'get_current_week',
 'load_combine',
 'load_contracts',
 'load_depth_charts',
 'load_draft_picks',
 'load_ff_opportunity',
 'load_ff_playerids',
 'load_ff_rankings',
 'load_ffverse',
 'load_ftn_charting',
 'load_injuries',
 'load_nextgen_stats',
 'load_officials',
 'load_participation',
 'load_pbp',
 'load_pfr_advstats',
 'load_player_stats',
 'load_players',
 'load_rosters',
 'load_rosters_weekly',
 'load_schedules',
 'load_snap_counts',
 'load_stats',
 'load_team_stats',
 'load_teams',
 'load_trades',
 'utils_date',
 'version']

In [4]:
seasons = list(range(2012, 2026))

player_stats = nfl.load_player_stats(
    seasons=seasons,
    summary_level="reg"
)

players = nfl.load_players()
contracts = nfl.load_contracts()

In [5]:
print("player Statistics:", player_stats.shape)
print("Players:", players.shape)
print("Contracts:', contacts.shape")

player Statistics: (26830, 143)
Players: (25036, 39)
Contracts:', contacts.shape


In [6]:
player_stats.columns

['player_id',
 'player_name',
 'player_display_name',
 'position',
 'position_group',
 'headshot_url',
 'season',
 'season_type',
 'recent_team',
 'games',
 'completions',
 'attempts',
 'passing_yards',
 'passing_tds',
 'passing_interceptions',
 'sacks_suffered',
 'sack_yards_lost',
 'sack_fumbles',
 'sack_fumbles_lost',
 'passing_air_yards',
 'passing_yards_after_catch',
 'passing_first_downs',
 'passing_epa',
 'passing_cpoe',
 'passing_2pt_conversions',
 'pacr',
 'passing_10',
 'passing_16',
 'passing_20',
 'passing_40',
 'carries',
 'rushing_yards',
 'rushing_tds',
 'rushing_fumbles',
 'rushing_fumbles_lost',
 'rushing_first_downs',
 'rushing_epa',
 'rushing_2pt_conversions',
 'rushing_10',
 'rushing_12',
 'rushing_20',
 'rushing_40',
 'receptions',
 'targets',
 'receiving_yards',
 'receiving_tds',
 'receiving_fumbles',
 'receiving_fumbles_lost',
 'receiving_air_yards',
 'receiving_yards_after_catch',
 'receiving_first_downs',
 'receiving_epa',
 'receiving_2pt_conversions',
 'receiv

In [7]:
players.columns

['gsis_id',
 'display_name',
 'common_first_name',
 'first_name',
 'last_name',
 'short_name',
 'football_name',
 'suffix',
 'esb_id',
 'nfl_id',
 'pfr_id',
 'pff_id',
 'otc_id',
 'espn_id',
 'smart_id',
 'birth_date',
 'position_group',
 'position',
 'ngs_position_group',
 'ngs_position',
 'height',
 'weight',
 'headshot',
 'college_name',
 'college_conference',
 'jersey_number',
 'rookie_season',
 'last_season',
 'latest_team',
 'status',
 'ngs_status',
 'ngs_status_short_description',
 'years_of_experience',
 'pff_position',
 'pff_status',
 'draft_year',
 'draft_round',
 'draft_pick',
 'draft_team']

In [8]:
contracts.columns

['player',
 'position',
 'team',
 'is_active',
 'year_signed',
 'years',
 'value',
 'apy',
 'guaranteed',
 'apy_cap_pct',
 'inflated_value',
 'inflated_apy',
 'inflated_guaranteed',
 'player_page',
 'otc_id',
 'gsis_id',
 'date_of_birth',
 'height',
 'weight',
 'college',
 'draft_year',
 'draft_round',
 'draft_overall',
 'draft_team',
 'cols']

In [9]:
player_stats.head()

player_id,player_name,player_display_name,position,position_group,headshot_url,season,season_type,recent_team,games,completions,attempts,passing_yards,passing_tds,passing_interceptions,sacks_suffered,sack_yards_lost,sack_fumbles,sack_fumbles_lost,passing_air_yards,passing_yards_after_catch,passing_first_downs,passing_epa,passing_cpoe,passing_2pt_conversions,pacr,passing_10,passing_16,passing_20,passing_40,carries,rushing_yards,rushing_tds,rushing_fumbles,rushing_fumbles_lost,rushing_first_downs,rushing_epa,…,fg_missed_0_19,fg_missed_20_29,fg_missed_30_39,fg_missed_40_49,fg_missed_50_59,fg_missed_60_,fg_made_list,fg_missed_list,fg_blocked_list,fg_made_distance,fg_missed_distance,fg_blocked_distance,pat_made,pat_att,pat_missed,pat_blocked,pat_pct,gwfg_made,gwfg_att,gwfg_missed,gwfg_blocked,gwfg_distance_list,pt_att,pt_blocked,pt_long,pt_yards,pt_inside_20,pt_out_of_bounds,pt_downed,pt_touchback,pt_fair_caught,pt_returned,pt_return_yards,pt_return_tds,pt_net_yards,fantasy_points,fantasy_points_ppr
str,str,str,str,str,str,i32,str,str,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,f64,f64,i32,f64,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,f64,…,i32,i32,i32,i32,i32,i32,str,str,str,i32,i32,i32,i32,i32,i32,i32,f64,i32,i32,i32,i32,str,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,f64,f64
"""00-0000108""","""D.Akers""","""David Akers""","""K""","""SPEC""","""https://static.www.nfl.com/ima…",2012,"""REG""","""SF""",16,0,0,0,0,0,0,0,0,0,0,0,0,null,null,0,null,0,0,0,0,0,0,0,0,0,0,null,…,0,0,1,5,4,0,"""40;43;63;36;48;29;29;36;40;19;…","""55;40;43;52;41;50;51;39;44;40""","""43;33;21""",1022,455,97,44,44,0,0,1.0,0,0,0,0,null,0,0,null,0,0,0,0,0,0,0,0,0,0,0.0,0.0
"""00-0000551""","""B.Ayanbadejo""","""Brendon Ayanbadejo""","""ILB""","""LB""","""https://static.www.nfl.com/ima…",2012,"""REG""","""BAL""",13,0,0,0,0,0,0,0,0,0,0,0,0,null,null,0,null,0,0,0,0,0,0,0,0,0,0,null,…,0,0,0,0,0,0,null,null,null,0,0,0,0,0,0,0,null,0,0,0,0,null,0,0,null,0,0,0,0,0,0,0,0,0,0,0.0,0.0
"""00-0000585""","""C.Bailey""","""Champ Bailey""","""CB""","""DB""","""https://static.www.nfl.com/ima…",2012,"""REG""","""DEN""",16,0,0,0,0,0,0,0,0,0,0,0,0,null,null,0,null,0,0,0,0,0,0,0,0,0,0,null,…,0,0,0,0,0,0,null,null,null,0,0,0,0,0,0,0,null,0,0,0,0,null,0,0,null,0,0,0,0,0,0,0,0,0,0,0.0,0.0
"""00-0000741""","""R.Barber""","""Ronde Barber""","""FS""","""DB""","""https://static.www.nfl.com/ima…",2012,"""REG""","""TB""",16,0,0,0,0,0,0,0,0,0,0,0,0,null,null,0,null,0,0,0,0,0,0,0,0,0,0,null,…,0,0,0,0,0,0,null,null,null,0,0,0,0,0,0,0,null,0,0,0,0,null,0,0,null,0,0,0,0,0,0,0,0,0,0,0.0,0.0
"""00-0000865""","""C.Batch""","""Charlie Batch""","""QB""","""QB""","""https://static.www.nfl.com/ima…",2012,"""REG""","""PIT""",2,45,70,475,1,4,3,-12,0,0,663,257,22,-1.646032,1.449535,0,0.71644,23,8,5,1,0,0,0,0,0,0,null,…,0,0,0,0,0,0,null,null,null,0,0,0,0,0,0,0,null,0,0,0,0,null,0,0,null,0,0,0,0,0,0,0,0,0,0,15.0,15.0


In [10]:
player_stats.shape

(26830, 143)

In [11]:
print("=== PLAYER STATISTICS ===")
print(f"Rows, Columns: {player_stats.shape}")
print("\nData Types:")
print(player_stats.dtypes)
print("\nMissing Values:")
print(player_stats.null_count())

=== PLAYER STATISTICS ===
Rows, Columns: (26830, 143)

Data Types:
[String, String, String, String, String, String, Int32, String, String, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Float64, Float64, Int32, Float64, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Float64, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Float64, Int32, Int32, Int32, Int32, Int32, Float64, Float64, Float64, Float64, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Float64, Float64, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Float64, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, String, String, String, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Float64, Int32, Int32, Int32, Int32, String

In [12]:
print("=== PLAYERS ===")
print(f"Rows, Columns: {players.shape}")
print("\nData Types:")
print(players.dtypes)
print("\nMissing Values:")
print(players.null_count())

=== PLAYERS ===
Rows, Columns: (25036, 39)

Data Types:
[String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, Int32, Int32, String, String, String, String, Int32, Int32, String, String, String, String, Int32, String, String, Int32, Int32, Int32, String]

Missing Values:
shape: (1, 39)
┌─────────┬────────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬───────────┐
│ gsis_id ┆ display_na ┆ common_fi ┆ first_nam ┆ … ┆ draft_yea ┆ draft_rou ┆ draft_pic ┆ draft_tea │
│ ---     ┆ me         ┆ rst_name  ┆ e         ┆   ┆ r         ┆ nd        ┆ k         ┆ m         │
│ u32     ┆ ---        ┆ ---       ┆ ---       ┆   ┆ ---       ┆ ---       ┆ ---       ┆ ---       │
│         ┆ u32        ┆ u32       ┆ u32       ┆   ┆ u32       ┆ u32       ┆ u32       ┆ u32       │
╞═════════╪════════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡
│

In [13]:
print("=== CONTRACTS ===")
print(f"Rows, Columns: {contracts.shape}")
print("\nData Types:")
print(contracts.dtypes)
print("\nMissing Values:")
print(contracts.null_count())

=== CONTRACTS ===
Rows, Columns: (51723, 25)

Data Types:
[String, String, String, Boolean, Int32, Int32, Float64, Float64, Float64, Float64, Float64, Float64, Float64, String, Int32, String, String, String, String, String, Int32, Int32, Int32, String, List(Struct({'year': String, 'team': String, 'base_salary': Float64, 'prorated_bonus': Float64, 'roster_bonus': Float64, 'guaranteed_salary': Float64, 'cap_number': Float64, 'cap_percent': Float64, 'cash_paid': Float64, 'workout_bonus': Float64, 'other_bonus': Float64, 'per_game_roster_bonus': Float64, 'option_bonus': Float64}))]

Missing Values:
shape: (1, 25)
┌────────┬──────────┬──────┬───────────┬───┬─────────────┬───────────────┬────────────┬──────┐
│ player ┆ position ┆ team ┆ is_active ┆ … ┆ draft_round ┆ draft_overall ┆ draft_team ┆ cols │
│ ---    ┆ ---      ┆ ---  ┆ ---       ┆   ┆ ---         ┆ ---           ┆ ---        ┆ ---  │
│ u32    ┆ u32      ┆ u32  ┆ u32       ┆   ┆ u32         ┆ u32           ┆ u32        ┆ u32  │
╞══

In [14]:
for i, col in enumerate(player_stats.columns, start=1):
    print(f"{i:3}. {col}")

  1. player_id
  2. player_name
  3. player_display_name
  4. position
  5. position_group
  6. headshot_url
  7. season
  8. season_type
  9. recent_team
 10. games
 11. completions
 12. attempts
 13. passing_yards
 14. passing_tds
 15. passing_interceptions
 16. sacks_suffered
 17. sack_yards_lost
 18. sack_fumbles
 19. sack_fumbles_lost
 20. passing_air_yards
 21. passing_yards_after_catch
 22. passing_first_downs
 23. passing_epa
 24. passing_cpoe
 25. passing_2pt_conversions
 26. pacr
 27. passing_10
 28. passing_16
 29. passing_20
 30. passing_40
 31. carries
 32. rushing_yards
 33. rushing_tds
 34. rushing_fumbles
 35. rushing_fumbles_lost
 36. rushing_first_downs
 37. rushing_epa
 38. rushing_2pt_conversions
 39. rushing_10
 40. rushing_12
 41. rushing_20
 42. rushing_40
 43. receptions
 44. targets
 45. receiving_yards
 46. receiving_tds
 47. receiving_fumbles
 48. receiving_fumbles_lost
 49. receiving_air_yards
 50. receiving_yards_after_catch
 51. receiving_first_downs
 52. 

In [15]:
for i, col in enumerate(players.columns, start=1):
    print(f"{i:3}. {col}")

  1. gsis_id
  2. display_name
  3. common_first_name
  4. first_name
  5. last_name
  6. short_name
  7. football_name
  8. suffix
  9. esb_id
 10. nfl_id
 11. pfr_id
 12. pff_id
 13. otc_id
 14. espn_id
 15. smart_id
 16. birth_date
 17. position_group
 18. position
 19. ngs_position_group
 20. ngs_position
 21. height
 22. weight
 23. headshot
 24. college_name
 25. college_conference
 26. jersey_number
 27. rookie_season
 28. last_season
 29. latest_team
 30. status
 31. ngs_status
 32. ngs_status_short_description
 33. years_of_experience
 34. pff_position
 35. pff_status
 36. draft_year
 37. draft_round
 38. draft_pick
 39. draft_team


In [16]:
for i, col in enumerate(contracts.columns, start=1):
    print(f"{i:3}. {col}")

  1. player
  2. position
  3. team
  4. is_active
  5. year_signed
  6. years
  7. value
  8. apy
  9. guaranteed
 10. apy_cap_pct
 11. inflated_value
 12. inflated_apy
 13. inflated_guaranteed
 14. player_page
 15. otc_id
 16. gsis_id
 17. date_of_birth
 18. height
 19. weight
 20. college
 21. draft_year
 22. draft_round
 23. draft_overall
 24. draft_team
 25. cols


In [17]:
print("Player Stats Key:")
print(player_stats.select("player_id").head())

Player Stats Key:
shape: (5, 1)
┌────────────┐
│ player_id  │
│ ---        │
│ str        │
╞════════════╡
│ 00-0000108 │
│ 00-0000551 │
│ 00-0000585 │
│ 00-0000741 │
│ 00-0000865 │
└────────────┘


In [18]:
[c for c in player_stats.columns if "id" in c.lower()]

['player_id', 'pt_inside_20']

In [19]:
[c for c in players.columns if "id" in c.lower()]

['gsis_id',
 'esb_id',
 'nfl_id',
 'pfr_id',
 'pff_id',
 'otc_id',
 'espn_id',
 'smart_id']

In [20]:
[c for c in contracts.columns if "id" in c.lower()]

['otc_id', 'gsis_id']

In [21]:
players.select([
    "gsis_id",
    "display_name"
]).head(10)

gsis_id,display_name
str,str
"""00-0028830""","""Isaako Aaitui"""
"""00-0038389""","""Israel Abanikanda"""
"""00-0024644""","""Jon Abbate"""
"""ABB498348""","""Vince Abbott"""
"""00-0031021""","""Jared Abbrederis"""
"""00-0032860""","""Mehdi Abdesmad"""
"""00-0028564""","""Isa Abdul-Quddus"""
"""00-0032104""","""Ameer Abdullah"""
"""00-0023663""","""Hamza Abdullah"""


In [22]:
player_stats.select([
    "player_id",
    "player_name"
]).head(10)

player_id,player_name
str,str
"""00-0000108""","""D.Akers"""
"""00-0000551""","""B.Ayanbadejo"""
"""00-0000585""","""C.Bailey"""
"""00-0000741""","""R.Barber"""
"""00-0000865""","""C.Batch"""
"""00-0001263""","""M.Birk"""
"""00-0001820""","""K.Brooking"""
"""00-0004091""","""P.Dawson"""
"""00-0004541""","""D.Driver"""


In [23]:
players.filter(
    pl.col("display_name").str.contains("Aaron Rodgers", literal=False)
).select(["display_name", "gsis_id", "esb_id", "nfl_id"])

display_name,gsis_id,esb_id,nfl_id
str,str,str,str
"""Aaron Rodgers""","""00-0023459""","""ROD339293""","""29851"""


In [24]:
player_stats.filter(
    pl.col("player_name").str.contains("Aaron Rodgers", literal=False)
).select(["player_name", "player_id"])

player_name,player_id
str,str


In [25]:
nfl_master = (
    player_stats
    .join(
        players,
        left_on="player_id",
        right_on="gsis_id",
        how="left",
        suffix="_player"
    )
    .join(
        contracts,
        left_on="player_id",
        right_on="gsis_id",
        how="left",
        suffix="_contract"
    )
)

In [26]:
print("Player Stats Shape:", player_stats.shape)
print("Master Dataset Shape:", nfl_master.shape)
print("Unique Players:", nfl_master["player_id"].n_unique())

nfl_master.head()

Player Stats Shape: (26830, 143)
Master Dataset Shape: (167229, 205)
Unique Players: 7002


player_id,player_name,player_display_name,position,position_group,headshot_url,season,season_type,recent_team,games,completions,attempts,passing_yards,passing_tds,passing_interceptions,sacks_suffered,sack_yards_lost,sack_fumbles,sack_fumbles_lost,passing_air_yards,passing_yards_after_catch,passing_first_downs,passing_epa,passing_cpoe,passing_2pt_conversions,pacr,passing_10,passing_16,passing_20,passing_40,carries,rushing_yards,rushing_tds,rushing_fumbles,rushing_fumbles_lost,rushing_first_downs,rushing_epa,…,rookie_season,last_season,latest_team,status,ngs_status,ngs_status_short_description,years_of_experience,pff_position,pff_status,draft_year,draft_round,draft_pick,draft_team,player,position_contract,team,is_active,year_signed,years,value,apy,guaranteed,apy_cap_pct,inflated_value,inflated_apy,inflated_guaranteed,player_page,otc_id_contract,date_of_birth,height_contract,weight_contract,college,draft_year_contract,draft_round_contract,draft_overall,draft_team_contract,cols
str,str,str,str,str,str,i32,str,str,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,f64,f64,i32,f64,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,f64,…,i32,i32,str,str,str,str,i32,str,str,i32,i32,i32,str,str,str,str,bool,i32,i32,f64,f64,f64,f64,f64,f64,f64,str,i32,str,str,str,str,i32,i32,i32,str,list[struct[13]]
"""00-0000108""","""D.Akers""","""David Akers""","""K""","""SPEC""","""https://static.www.nfl.com/ima…",2012,"""REG""","""SF""",16,0,0,0,0,0,0,0,0,0,0,0,0,null,null,0,null,0,0,0,0,0,0,0,0,0,0,null,…,1998,2013,"""DET""","""ACT""",null,null,16,"""K""",null,null,null,null,null,"""David Akers""","""K""","""Lions""",false,0,1,1.005,1.005,0.1,0.0,0.0,0.0,0.0,"""https://overthecap.com/player/…",1272,"""December 9, 1974""","""5'10""""","""200""","""Louisville""",1997,null,null,"""Panthers""","[{""2013"",""Lions"",0.94,0.065,null,null,0.62,0.005,1.005,null,null,null,null}, {""Total"",""Total"",0.94,0.065,null,null,0.62,null,1.005,null,null,null,null}]"
"""00-0000551""","""B.Ayanbadejo""","""Brendon Ayanbadejo""","""ILB""","""LB""","""https://static.www.nfl.com/ima…",2012,"""REG""","""BAL""",13,0,0,0,0,0,0,0,0,0,0,0,0,null,null,0,null,0,0,0,0,0,0,0,0,0,0,null,…,2003,2012,"""BAL""","""ACT""",null,null,10,"""LB""",null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null
"""00-0000585""","""C.Bailey""","""Champ Bailey""","""CB""","""DB""","""https://static.www.nfl.com/ima…",2012,"""REG""","""DEN""",16,0,0,0,0,0,0,0,0,0,0,0,0,null,null,0,null,0,0,0,0,0,0,0,0,0,0,null,…,1999,2013,"""DEN""","""ACT""",null,null,15,"""CB""",null,1999,1,7,"""WAS""","""Champ Bailey""","""CB""","""Broncos""",false,2004,7,63.032025,9.004575,6.0,0.112,235.601573,33.657368,22.426845,"""https://overthecap.com/player/…",4,"""June 22, 1978""","""6'0""""","""192""","""Georgia""",1999,1,7,"""Commanders""","[{""1999"",""Redskins"",0.592,1.04,0.0,0.0,1.632,0.028,5.792,0.0,null,null,null}, {""2000"",""Redskins"",0.275,1.1465,0.0,0.0,1.5215,0.024,0.801,0.1,null,null,null}, … {""Total"",""Total"",58.969,30.831,14.032025,18.0,107.632025,null,107.632025,3.8,null,null,null}]"
"""00-0000585""","""C.Bailey""","""Champ Bailey""","""CB""","""DB""","""https://static.www.nfl.com/ima…",2012,"""REG""","""DEN""",16,0,0,0,0,0,0,0,0,0,0,0,0,null,null,0,null,0,0,0,0,0,0,0,0,0,0,null,…,1999,2013,"""DEN""","""ACT""",null,null,15,"""CB""",null,1999,1,7,"""WAS""","""Champ Bailey""","""CB""","""Broncos""",false,2011,4,43.0,10.75,10.5,0.089,107.593769,26.898442,26.272897,"""https://overthecap.com/player/…",4,"""June 22, 1978""","""6'0""""","""192""","""Georgia""",1999,1,7,"""Commanders""","[{""1999"",""Redskins"",0.592,1.04,0.0,0.0,1.632,0.028,5.792,0.0,null,null,null}, {""2000"",""Redskins"",0.275,1.1465,0.0,0.0,1.5215,0.024,0.801,0.1,null,null,null}, … {""Total"",""Total"",58.969,30.831,14.032025,18.0,107.632025,null,107.632025,3.8,null,null,null}]"
"""00-0000585""","""C.Bailey""","""Champ Bailey""","""CB""","""DB""","""https://stati

In [27]:
join_validation = nfl_master.select([
    pl.len().alias("total_rows"),
    pl.col("display_name").is_null().sum().alias("missing_player_matches"),
    pl.col("player").is_null().sum().alias("missing_contract_matches")
])

join_validation

total_rows,missing_player_matches,missing_contract_matches
u32,u32,u32
167229,14,472


## 7. Contract Duplication

The merged dataset contains more rows than the player statistics dataset. The increase is caused by multiple contract records for individual players. 

In [28]:
contract_counts = (
    contracts
    .group_by("gsis_id")
    .len()
    .sort("len", descending=True)
)

contract_counts.head(20)

gsis_id,len
str,u32
null,4190
"""00-0032463""",38
"""00-0035406""",37
"""00-0032891""",34
"""00-0032951""",34
…,…
"""00-0026300""",29
"""00-0034966""",29
"""00-0035189""",29


In [29]:
# Count missing GSIS IDs in the contracts table

null_gsis = contracts.filter(pl.col("gsis_id").is_null())

print(f"Contract records: {contracts.height:,}")
print(f"Missing gsis_id: {null_gsis.height:,}")
print(f"Percent missing: {100 * null_gsis.height / contracts.height:.2f}%")

Contract records: 51,723
Missing gsis_id: 4,190
Percent missing: 8.10%


In [30]:
# Remove records without a GSIS ID and examine duplicates again

contracts_valid = contracts.filter(pl.col("gsis_id").is_not_null())

contract_counts = (
    contracts_valid
    .group_by("gsis_id")
    .len()
    .sort("len", descending=True)
)

contract_counts.head(20)

gsis_id,len
str,u32
"""00-0032463""",38
"""00-0035406""",37
"""00-0032951""",34
"""00-0032891""",34
"""00-0034461""",33
…,…
"""00-0035363""",29
"""00-0034966""",29
"""00-0034587""",28


In [31]:
# Count contract records with missing gsis ID's

null_gsis = contracts.filter(
    pl.col("gsis_id").is_null()
)

print(f"Total contract records: {contracts.height:,}")
print(f"missing gsis_id: {null_gsis.height:,}")
print(
    f"Percent missing: "f"{100 * null_gsis.height / contracts.height:.2f}%"
)

Total contract records: 51,723
missing gsis_id: 4,190
Percent missing: 8.10%


In [32]:
# Retain contract records with a valid GSIS ID

contracts_valid = contracts.filter(
    pl.col("gsis_id").is_not_null()
)

# Count contract records for each identified player

contract_counts_valid = (
    contracts_valid
    .group_by("gsis_id")
    .len(name="contract_record_count")
    .sort("contract_record_count", descending=True)
)

contract_counts_valid.head(20)

gsis_id,contract_record_count
str,u32
"""00-0032463""",38
"""00-0035406""",37
"""00-0032891""",34
"""00-0032951""",34
"""00-0034461""",33
…,…
"""00-0035363""",29
"""00-0035189""",29
"""00-0030294""",28


In [33]:
# Display all columns in the contracts dataset

print(f"Number of columns: {len(contracts.columns)}")

for column in contracts.columns:
    print(column)

Number of columns: 25
player
position
team
is_active
year_signed
years
value
apy
guaranteed
apy_cap_pct
inflated_value
inflated_apy
inflated_guaranteed
player_page
otc_id
gsis_id
date_of_birth
height
weight
college
draft_year
draft_round
draft_overall
draft_team
cols


In [34]:
# Display every contract record for one player

sample_player = "00-0032463"

(
    contracts_valid
    .filter(pl.col("gsis_id") == sample_player)
    .sort("year_signed")
)

player,position,team,is_active,year_signed,years,value,apy,guaranteed,apy_cap_pct,inflated_value,inflated_apy,inflated_guaranteed,player_page,otc_id,gsis_id,date_of_birth,height,weight,college,draft_year,draft_round,draft_overall,draft_team,cols
str,str,str,bool,i32,i32,f64,f64,f64,f64,f64,f64,f64,str,i32,str,str,str,str,str,i32,i32,i32,str,list[struct[13]]
"""Kyle Peko""","""IDL""","""Broncos""",false,2016,2,0.99,0.495,0.0,0.003,1.920448,0.960224,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""July 23, 1993""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","[{""2016"",""Broncos"",0.45,null,null,null,0.026471,0.0,0.026471,null,null,null,null}, {""2017"",""Broncos"",0.54,null,null,null,0.190588,0.001,0.190588,null,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,2.604706,null,2.92665,null,null,null,null}]"
"""Kyle Peko""","""IDL""","""Broncos""",false,2017,2,1.17,0.585,0.0,0.004,2.110204,1.055102,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""July 23, 1993""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","[{""2016"",""Broncos"",0.45,null,null,null,0.026471,0.0,0.026471,null,null,null,null}, {""2017"",""Broncos"",0.54,null,null,null,0.190588,0.001,0.190588,null,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,2.604706,null,2.92665,null,null,null,null}]"
"""Kyle Peko""","""IDL""","""Broncos""",false,2017,1,0.54,0.54,0.0,0.003,0.97394,0.97394,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""July 23, 1993""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","[{""2016"",""Broncos"",0.45,null,null,null,0.026471,0.0,0.026471,null,null,null,null}, {""2017"",""Broncos"",0.54,null,null,null,0.190588,0.001,0.190588,null,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,2.604706,null,2.92665,null,null,null,null}]"
"""Kyle Peko""","""IDL""","""Broncos""",false,2017,1,0.54,0.54,0.0,0.003,0.97394,0.97394,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""July 23, 1993""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","[{""2016"",""Broncos"",0.45,null,null,null,0.026471,0.0,0.026471,null,null,null,null}, {""2017"",""Broncos"",0.54,null,null,null,0.190588,0.001,0.190588,null,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,2.604706,null,2.92665,null,null,null,null}]"
"""Kyle Peko""","""IDL""","""Broncos""",false,2017,1,0.1224,0.1224,0.0,0.001,0.22076,0.22076,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""July 23, 1993""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","[{""2016"",""Broncos"",0.45,null,null,null,0.026471,0.0,0.026471,null,null,null,null}, {""2017"",""Broncos"",0.54,null,null,null,0.190588,0.001,0.190588,null,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,2.604706,null,2.92665,null,null,null,null}]"
…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…
"""Kyle Peko""","""IDL""","""Lions""",false,2024,1,0.352822,0.352822,0.0,0.001,0.416092,0.416092,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""July 23, 1993""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","[{""2016"",""Broncos"",0.45,null,null,null,0.026471,0.0,0.026471,null,null,null,null}, {""2017"",""Broncos"",0.54,null,null,null,0.190588,0.001,0.190588,null,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,2.604706,null,2.92665,null,null,null,null}]"
"""Kyle Peko""","""IDL""","""Lions""",false,2024,1,0.3024,0.3024,0.0,0.001,0.356628,0.356628,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""July 23, 1993""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","[{""2016"",""Broncos"",0.45,null,null,null,0.026471,0.0,0.026471,null,null,null,null}, {""2017"",""Broncos"",0.54,null,null,null,0.190588,0.001,0.190588,null,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,2.604706,null,2.92665,null,null,null,null}]"
"""Kyle Peko""","""

In [35]:
# Count exact duplicate rows in the contracts dataset

contracts_unique = contracts.unique()

duplicate_count = contracts.height - contracts_unique.height

print(f"Original rows: {contracts.height:,}")
print(f"Unique rows: {contracts_unique.height:,}")
print(f"Exact duplicate rows: {duplicate_count:,}")

Original rows: 51,723
Unique rows: 48,393
Exact duplicate rows: 3,330


# Current findings: 

Missing gsis_id	4,190 rows (8.10%)	Cannot be joined to player statistics
Maximum contracts per player  38	Players legitimately have multiple contract records
Exact duplicate rows	3,330	    The contracts dataset contains duplicate records

In [36]:
# Select one player and remove the nested "cols" column for easier comparison

sample_player = "00-0032463"

(
    contracts_valid
    .filter(pl.col("gsis_id") == sample_player)
    .drop("cols")
    .sort(["year_signed", "team", "value"])
)

player,position,team,is_active,year_signed,years,value,apy,guaranteed,apy_cap_pct,inflated_value,inflated_apy,inflated_guaranteed,player_page,otc_id,gsis_id,date_of_birth,height,weight,college,draft_year,draft_round,draft_overall,draft_team
str,str,str,bool,i32,i32,f64,f64,f64,f64,f64,f64,f64,str,i32,str,str,str,str,str,i32,i32,i32,str
"""Kyle Peko""","""IDL""","""Broncos""",false,2016,2,0.99,0.495,0.0,0.003,1.920448,0.960224,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""July 23, 1993""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos"""
"""Kyle Peko""","""IDL""","""Broncos""",false,2017,1,0.1224,0.1224,0.0,0.001,0.22076,0.22076,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""July 23, 1993""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos"""
"""Kyle Peko""","""IDL""","""Broncos""",false,2017,1,0.1224,0.1224,0.0,0.001,0.22076,0.22076,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""July 23, 1993""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos"""
"""Kyle Peko""","""IDL""","""Broncos""",false,2017,1,0.54,0.54,0.0,0.003,0.97394,0.97394,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""July 23, 1993""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos"""
"""Kyle Peko""","""IDL""","""Broncos""",false,2017,1,0.54,0.54,0.0,0.003,0.97394,0.97394,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""July 23, 1993""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos"""
…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…
"""Kyle Peko""","""IDL""","""Lions""",false,2024,1,1.21,1.21,0.0,0.005,1.426985,1.426985,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""July 23, 1993""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos"""
"""Kyle Peko""","""IDL""","""Lions""",false,2024,1,1.21,1.21,0.0,0.005,1.426985,1.426985,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""July 23, 1993""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos"""
"""Kyle Peko""","""IDL""","""Chargers""",false,2025,1,0.315,0.315,0.0,0.001,0.339821,0.339821,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""July 23, 1993""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos"""


### Findings Summary

The contracts dataset was examined to understand the increase in rows following integration with the player statistics dataset.

findings:

- 4,190 contract records (8.10%) contain missing `gsis_id` values and cannot be linked to player statistics.
- Many players legitimately possess multiple contract records because they signed multiple contracts throughout their careers.
- The contracts dataset also contains exact duplicate records. A comparison of individual player histories confirmed that several contracts are duplicated without any differences across the available fields.
- The increased row count after merging is attributable to both, one-to-many relationships and duplicate records present in the source contracts dataset.

No records were modified or removed yet. Data cleaning decisions will be deferred revisited during preprocessing stage after EDA.

In [37]:
player_season_master = (
    player_stats
    .join(
        players,
        left_on="player_id",
        right_on="gsis_id",
        how="left"
    )
)

print(player_season_master.shape)

(26830, 181)


In [38]:
wr_master = (
    player_season_master
    .filter(pl.col("position") == "WR")
)

print(f"Wide receiver records: {wr_master.height:,}")

wr_master.select("position").unique()

Wide receiver records: 3,148


position
str
"""WR"""


In [39]:
print(player_season_master.shape)
print(wr_master.shape)

(26830, 181)
(3148, 181)


In [40]:
(
    player_season_master
    .select("season")
    .unique()
    .sort("season")
)

season
i32
2012
2013
2014
2015
2016
…
2021
2022
2023


In [41]:
(
    player_season_master
    .group_by(["player_id", "season"])
    .len()
    .filter(pl.col("len") > 1)
)

player_id,season,len
str,i32,u32


## Summary

In this notebook, the data collection and integration phase established the foundation for the predictive modeling workflow. Multiple datasets containing player statistics, player metadata, and contract information were loaded, inspected, and evaluated for quality and compatibility.

Key validation steps confirmed the structure of the player statistics data, verified unique player season observations, and examined identifier consistency across datasets. During integration, exploratory analysis revealed that the contract dataset contains multiple contract records for individual players, resulting in duplicate player season observations when merged directly. Rather than introducing duplicated records into the analytical dataset, the contract merge was deferred until a season aware integration strategy can be developed.

A clean master dataset was then created by combining player statistics with player metadata, followed by filtering to wide receivers (WR), the position selected for this study. Validation confirmed that the resulting dataset contains one observation per player per season with no duplicate player-season combinations.

The completed dataset provides a reliable foundation for exploratory data analysis (EDA), feature engineering, and predictive modeling in the subsequent notebooks.

### Deliverables
- Loaded all required datasets.
- Inspected dataset structure and schema.
- Validated player identifiers and season coverage.
- Investigated contract data integration challenges.
- Created the `player_season_master` dataset.
- Filtered the dataset to wide receivers.
- Verified dataset integrity and uniqueness.

In [42]:
from pathlib import Path

output_path = Path("../data/processed/wr_master.csv")
output_path.parent.mkdir(parents=True, exist_ok=True)

wr_master.write_csv(output_path)

print("Export complete!")
print(f"Location: {output_path.resolve()}")
print(f"Rows: {wr_master.height:,}")
print(f"Columns: {wr_master.width}")

Export complete!
Location: C:\Users\billf\Capstone-Project\data\processed\wr_master.csv
Rows: 3,148
Columns: 181
